jupyterhub/handlers/base.py

```diff
@@ -336,11 +336,15 @@ async def refresh_auth(self, user, force=False):
 
     def get_token(self):
         """get token from authorization header"""
-        token = self.get_auth_token()
-        if token is None:
-            return None
-        orm_token = orm.APIToken.find(self.db, token)
-        return orm_token
+        try:
+            token = self.get_auth_token()
+            if token is None:
+                return None
+            orm_token = orm.APIToken.find(self.db, token)
+            return orm_token
+        except SQLAlchemyError:
+            self.log.exception("Rolling back session due to database error")
+            self.db.rollback()
```

Comment:
This one is surprising, because we already have this catch for unhandled 500 errors. Do you have a sample log snippet with some context for the error occurring here?

Reply:
Below is one sample log. This particular try/except has shown up a number of times for us in the last month:

[I 2021-08-03 19:03:18.365 JupyterHub log:181] 200 GET /hub/metrics (prometheus@10.78.157.41) 9.07ms
[W 2021-08-03 19:03:17.139 JupyterHub log:181] 403 GET /hub/metrics (@10.78.37.51) 4.05ms
UnboundLocalError: local variable 'orm_token' referenced before assignment
recorded = self._record_activity(orm_token, now)
File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 367, in get_current_user_token
user = self.get_current_user_token()
File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 423, in get_current_user
Traceback (most recent call last):
[E 2021-08-03 19:03:17.138 JupyterHub base:432] Error getting current user
(Background on this error at: http://sqlalche.me/e/13/e3q8)
[parameters: {'prefix': '59f4', 'expires_at_1': datetime.datetime(2021, 8, 3, 19, 3, 17, 136102)}]
WHERE (%(prefix)s LIKE api_tokens.prefix || '%%') AND (api_tokens.expires_at IS NULL OR api_tokens.expires_at >= %(expires_at_1)s)]
FROM api_tokens
[SQL: SELECT api_tokens.user_id AS api_tokens_user_id, api_tokens.service_id AS api_tokens_service_id, api_tokens.id AS api_tokens_id, api_tokens.hashed AS api_tokens_hashed, api_tokens.prefix AS api_tokens_prefix, api_tokens.created AS api_tokens_created, api_tokens.expires_at AS api_tokens_expires_at, api_tokens.last_activity AS api_tokens_last_activity, api_tokens.note AS api_tokens_note
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) SSL connection has been closed unexpectedly
cursor.execute(statement, parameters)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/default.py", line 593, in do_execute
self.dialect.do_execute(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
raise exception
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/util/compat.py", line 182, in raise_
util.raise_(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_exception
self._handle_dbapi_exception(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
ret = self._execute_context(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1124, in _execute_clauseelement
return connection._execute_clauseelement(self, multiparams, params)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
return meth(self, multiparams, params)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1011, in execute
result = conn.execute(querycontext.statement, self._params)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3560, in _execute_and_instances
return self._execute_and_instances(context)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3535, in __iter__
for orm_token in prefix_match:
File "/usr/local/lib/python3.8/dist-packages/jupyterhub/orm.py", line 497, in find
orm_token = orm.APIToken.find(self.db, token)
File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 358, in get_current_user_token
Traceback (most recent call last):
[E 2021-08-03 19:03:17.137 JupyterHub base:362] Rolling back session due to database error testing123.
[I 2021-08-03 19:03:15.764 JupyterHub log:181] 200 GET /hub/metrics (prometheus@10.78.158.102) 9.08ms

Reply:
Gotcha, thanks! I worry a bit about adding the same try/catch to individual db operations, because it really should be at a high level on *all* operations in one (or a few) places. If it's not being caught for this one, it could occur in any of the dozens of other db operations as well. My guess is this one is common because it's typically the first one that happens in a request, but it could happen anywhere.


We have the catch-and-rollback [here](https://github.com/jupyterhub/jupyterhub/blob/1a1a60b02b592c23a42bb688cdd7fb31c9c8fac2/jupyterhub/handlers/base.py#L1274-L1276) in write_error (also [here](https://github.com/jupyterhub/jupyterhub/blob/1a1a60b02b592c23a42bb688cdd7fb31c9c8fac2/jupyterhub/apihandlers/base.py#L106-L116) for APIHandlers). That's what is *supposed* to trigger this same logic in this same event for any db operation anywhere during the course of the request. However, it's clearly not occurring.

Do you have a log surrounding this error when it occurred but *wasn't* rolled back? And what version fo jupyterhub is this?

Reply:
We are currently using version 1.2.1, but the below log is from version 1.1.0. This one looks to first show an error in `update_last_activity` in app.py, but then gets stuck in `get_current_user` in base.py. I have seen logs where the first error shows up in base.py, but I cannot find a log for that one. In the below scenario, for JupyterHub to recover we usually have to bounce the pod.
Full log file:
[hublog.txt](https://github.com/jupyterhub/jupyterhub/files/6938617/hublog.txt)


[I 2021-02-09 15:41:40.044 JupyterHub log:174] 200 GET /hub/health (@10.78.38.7) 0.67ms
[E 2021-02-09 15:41:40.486 JupyterHub ioloop:763] Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fb08fff1cf8>>, <Task finished coro=<JupyterHub.update_last_activity() done, defined at /usr/local/lib/python3.6/dist-packages/jupyterhub/app.py:2428> exception=StatementError("(sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back",)>)
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/tornado/ioloop.py", line 743, in _run_callback
        ret = callback()
      File "/usr/local/lib/python3.6/dist-packages/tornado/ioloop.py", line 767, in _discard_future_result
        future.result()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/app.py", line 2445, in update_last_activity
        user = orm.User.find(self.db, route_data['user'])
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/orm.py", line 221, in find
        return db.query(cls).filter(cls.name == name).first()
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3298, in first
        ret = list(self[0:1])
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3076, in __getitem__
        return list(res)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3403, in __iter__
        return self._execute_and_instances(context)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3428, in _execute_and_instances
        result = conn.execute(querycontext.statement, self._params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 984, in execute
        return meth(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 293, in _execute_on_connection
        return connection._execute_clauseelement(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1103, in _execute_clauseelement
        distilled_params,
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1178, in _execute_context
        e, util.text_type(statement), parameters, None, None
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1482, in _handle_dbapi_exception
        sqlalchemy_exception, with_traceback=exc_info[2], from_=e
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
        raise exception
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1173, in _execute_context
        conn = self._revalidate_connection()
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 457, in _revalidate_connection
        "Can't reconnect until invalid "
    sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back
    [SQL: SELECT users.id AS users_id, users.name AS users_name, users.admin AS users_admin, users.created AS users_created, users.last_activity AS users_last_activity, users.cookie_id AS users_cookie_id, users.state AS users_state, users.encrypted_auth_state AS users_encrypted_auth_state 
    FROM users 
    WHERE users.name = %(name_1)s 
     LIMIT %(param_1)s]
    [parameters: [immutabledict({})]]
    
[I 2021-02-09 15:42:11.480 JupyterHub log:174] 200 GET /hub/login (@::ffff:10.78.38.75) 1.07ms
[I 2021-02-09 15:42:12.044 JupyterHub log:174] 200 GET /hub/health (@10.78.38.7) 0.70ms
[I 2021-02-09 15:42:14.044 JupyterHub log:174] 200 GET /hub/health (@10.78.38.7) 0.67ms
[E 2021-02-09 15:42:15.836 JupyterHub base:431] Error getting current user
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 422, in get_current_user
        user = self.get_current_user_token()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 360, in get_current_user_token
        orm_token = orm.APIToken.find(self.db, token)
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/orm.py", line 453, in find
        for orm_token in prefix_match:
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3403, in __iter__
        return self._execute_and_instances(context)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3428, in _execute_and_instances
        result = conn.execute(querycontext.statement, self._params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 984, in execute
        return meth(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 293, in _execute_on_connection
        return connection._execute_clauseelement(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1103, in _execute_clauseelement
        distilled_params,
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1178, in _execute_context
        e, util.text_type(statement), parameters, None, None
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1482, in _handle_dbapi_exception
        sqlalchemy_exception, with_traceback=exc_info[2], from_=e
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
        raise exception
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1173, in _execute_context
        conn = self._revalidate_connection()
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 457, in _revalidate_connection
        "Can't reconnect until invalid "
    sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back
    [SQL: SELECT api_tokens.user_id AS api_tokens_user_id, api_tokens.service_id AS api_tokens_service_id, api_tokens.id AS api_tokens_id, api_tokens.hashed AS api_tokens_hashed, api_tokens.prefix AS api_tokens_prefix, api_tokens.created AS api_tokens_created, api_tokens.expires_at AS api_tokens_expires_at, api_tokens.last_activity AS api_tokens_last_activity, api_tokens.note AS api_tokens_note 
    FROM api_tokens 
    WHERE (%(prefix)s LIKE api_tokens.prefix || '%%') AND (api_tokens.expires_at IS NULL OR api_tokens.expires_at >= %(expires_at_1)s)]
    [parameters: [immutabledict({})]]
    
[W 2021-02-09 15:42:15.837 JupyterHub log:174] 403 POST /hub/api/users/pa190603683/activity (@10.78.38.44) 2.32ms
[I 2021-02-09 15:42:15.880 JupyterHub log:174] 200 GET /hub/login (@::ffff:10.78.37.37) 1.12ms
[I 2021-02-09 15:42:16.044 JupyterHub log:174] 200 GET /hub/health (@10.78.38.7) 0.65ms
[E 2021-02-09 15:42:16.748 JupyterHub base:431] Error getting current user
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 422, in get_current_user
        user = self.get_current_user_token()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 360, in get_current_user_token
        orm_token = orm.APIToken.find(self.db, token)
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/orm.py", line 453, in find
        for orm_token in prefix_match:
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3403, in __iter__
        return self._execute_and_instances(context)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3428, in _execute_and_instances
        result = conn.execute(querycontext.statement, self._params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 984, in execute
        return meth(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 293, in _execute_on_connection
        return connection._execute_clauseelement(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1103, in _execute_clauseelement
        distilled_params,
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1178, in _execute_context
        e, util.text_type(statement), parameters, None, None
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1482, in _handle_dbapi_exception
        sqlalchemy_exception, with_traceback=exc_info[2], from_=e
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
        raise exception
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1173, in _execute_context
        conn = self._revalidate_connection()
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 457, in _revalidate_connection
        "Can't reconnect until invalid "
    sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back
    [SQL: SELECT api_tokens.user_id AS api_tokens_user_id, api_tokens.service_id AS api_tokens_service_id, api_tokens.id AS api_tokens_id, api_tokens.hashed AS api_tokens_hashed, api_tokens.prefix AS api_tokens_prefix, api_tokens.created AS api_tokens_created, api_tokens.expires_at AS api_tokens_expires_at, api_tokens.last_activity AS api_tokens_last_activity, api_tokens.note AS api_tokens_note 
    FROM api_tokens 
    WHERE (%(prefix)s LIKE api_tokens.prefix || '%%') AND (api_tokens.expires_at IS NULL OR api_tokens.expires_at >= %(expires_at_1)s)]
    [parameters: [immutabledict({})]]
    
[W 2021-02-09 15:42:16.749 JupyterHub log:174] 403 POST /hub/api/users/pa190603683/activity (@10.78.38.44) 2.31ms
[I 2021-02-09 15:42:18.045 JupyterHub log:174] 200 GET /hub/health (@10.78.38.7) 0.86ms
[I 2021-02-09 15:42:18.291 JupyterHub log:174] 200 GET /hub/login (@::ffff:10.78.39.70) 1.07ms
[E 2021-02-09 15:42:18.436 JupyterHub base:431] Error getting current user
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 422, in get_current_user
        user = self.get_current_user_token()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 360, in get_current_user_token
        orm_token = orm.APIToken.find(self.db, token)
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/orm.py", line 453, in find
        for orm_token in prefix_match:
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3403, in __iter__
        return self._execute_and_instances(context)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3428, in _execute_and_instances
        result = conn.execute(querycontext.statement, self._params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 984, in execute
        return meth(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 293, in _execute_on_connection
        return connection._execute_clauseelement(self, multiparams, params)
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1103, in _execute_clauseelement
        distilled_params,
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1178, in _execute_context
        e, util.text_type(statement), parameters, None, None
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1482, in _handle_dbapi_exception
        sqlalchemy_exception, with_traceback=exc_info[2], from_=e
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
        raise exception
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1173, in _execute_context
        conn = self._revalidate_connection()
      File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 457, in _revalidate_connection
        "Can't reconnect until invalid "
    sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back
    [SQL: SELECT api_tokens.user_id AS api_tokens_user_id, api_tokens.service_id AS api_tokens_service_id, api_tokens.id AS api_tokens_id, api_tokens.hashed AS api_tokens_hashed, api_tokens.prefix AS api_tokens_prefix, api_tokens.created AS api_tokens_created, api_tokens.expires_at AS api_tokens_expires_at, api_tokens.last_activity AS api_tokens_last_activity, api_tokens.note AS api_tokens_note 
    FROM api_tokens 
    WHERE (%(prefix)s LIKE api_tokens.prefix || '%%') AND (api_tokens.expires_at IS NULL OR api_tokens.expires_at >= %(expires_at_1)s)]
    [parameters: [immutabledict({})]]
    
[W 2021-02-09 15:42:18.437 JupyterHub log:174] 403 POST /hub/api/users/pa190603683/activity (@10.78.38.44) 2.26ms
[I 2021-02-09 15:42:20.045 JupyterHub log:174] 200 GET /hub/health (@10.78.38.7) 0.69ms